taskq/consumer.py

```diff
@@ -102,66 +102,97 @@ def execute_tasks(self):
     def fetch_due_tasks(self):
         # Multiple instances of taskq rely on select_for_update().
         # This mechanism will lock selected rows until the end of the transaction.
-        # We also fetch STATUS_RUNNING in case of previous inconsistent state.
-        due_tasks = Task.objects.filter(
-            Q(status=Task.STATUS_QUEUED) | Q(status=Task.STATUS_RUNNING),
-            due_at__lte=timezone.now()
-        ).select_for_update(skip_locked=True)
+
+        def fetched(task):
+            task.status = Task.STATUS_FETCHED
+            return task
+
+        with transaction.atomic():
+            due_tasks = [
+                fetched(task)
+                for task in Task.objects.filter(
+                    Q(status=Task.STATUS_QUEUED), due_at__lte=timezone.now()
+                ).select_for_update(skip_locked=True)
+            ]
+            Task.objects.bulk_update(due_tasks, fields=["status"])
```

Comment:
I find this a bit difficult to read. Also, I'm not sure how defining a function `fetched()` which does only one thing is simpler? How about:

```py
with transaction.atomic():
    due_tasks = Task.objects.filter(
        Q(status=Task.STATUS_QUEUED), due_at__lte=timezone.now()
    ).select_for_update(skip_locked=True)
    
    for task in due_tasks:
        task.status = Task.STATUS_FETCHED
    
    Task.objects.bulk_update(due_tasks, fields=["status"])
```

This feels more straightforward to me?

Reply:
we can't write it like that because due_tasks is a queryset, so i'm almost sur the code you propose will not work (or at least not work as expected), but I can rewrite my code as :
```
    due_tasks = list(
        Task.objects.filter(
            Q(status=Task.STATUS_QUEUED), due_at__lte=timezone.now()
        ).select_for_update(skip_locked=True)
    )
    
    for task in due_tasks:
        task.status = Task.STATUS_FETCHED
    
    Task.objects.bulk_update(due_tasks, fields=["status"])
```

Reply:
I was wrong ^^"